src/use_case/add_routine/AddRoutineInteractor.java

```diff
@@ -1,20 +1,35 @@
 package use_case.add_routine;
 
-import entity.User;
+import entity.Routine;
+import entity.RoutineFactory;
 
 public class AddRoutineInteractor implements AddRoutineInputBoundary{
-
-    final AddRoutineUserDataAccessInterface addRoutineUserDataAccessInterface;
     final AddRoutineOutputBoundary addRoutinePresenter;
+    final AddRoutineDataAccessInterface addRoutineDataAccessObject;
+    final RoutineFactory routineFactory;
+    private static int lastAssignedId = 0;
+
+    private synchronized int generateId() {
+        return ++lastAssignedId; // generate an id for each routine created
+    }
 
-    public AddRoutineInteractor(AddRoutineUserDataAccessInterface addRoutineUserDataAccessInterface,
-                           LoginOutputBoundary loginOutputBoundary) {
-        this.userDataAccessObject = userDataAccessInterface;
-        this.loginPresenter = loginOutputBoundary;
+    public AddRoutineInteractor(AddRoutineDataAccessInterface addRoutineDataAccessObject,
+                                AddRoutineOutputBoundary addRoutineOutputBoundary, RoutineFactory routineFactory) {
+        this.addRoutineDataAccessObject = addRoutineDataAccessObject;
+        this.addRoutinePresenter = addRoutineOutputBoundary;
+        this.routineFactory = routineFactory;
     }
 
     @Override
     public void execute(AddRoutineInputData addRoutineInputData) {
+        if (addRoutineDataAccessObject.existsByName(addRoutineInputData.getRoutineName())) {
+            addRoutinePresenter.prepareFailView("Routine name already exists.");
```

Comment:
if addRoutineDataAccessObject.existsByName(addRoutineInputData.getRoutineName()) is true, shouldn't it prepare success view and do the work here?